### PR TITLE
Products: Grouping design for product creation type sheet

### LINKED ISSUE #11361: Products: Improve the bottom sheet for product type selection
Currently, we have two separate bottom sheets when creating a new product if the user is eligible for product creation AI:

https://github.com/woocommerce/woocommerce-ios/assets/5533851/0ecc4d91-d4c6-4599-b3a9-c174b52a1139

We should enhance the experience by reuniting the options in one sheet with separators between product types:

https://github.com/woocommerce/woocommerce-ios/assets/5533851/35045074-af53-4a34-961d-043af0a322a1

Internal discussion: p1700008726769739-slack-C0354HSNUJH.

### PR DESCRIPTION
**Pull request by hafizrahman**
Closes: #11361, #11211

## Description

This PR is the second part of the product type creation bottom sheet.

Specifically, it now groups the "Manual product creation" into three categories:

1. Standard
   - Physical product
   - Virtual product
   - Variable product
2. Subscription
   - Simple subscription
   - Variable subscription
3. Other
   - Grouped product
   - External product

## Testing instructions

1. Go to Products tab,
2. Create new product,
3. Ensure that the bottom sheet demo matches the video below, and with the design video on https://github.com/woocommerce/woocommerce-ios/issues/11361#issue-2027620643
4. Additionally, test with a site that doesn't have subscriptions support. Ensure that the subscriptions category and its product types are not shown.

## Video

https://github.com/woocommerce/woocommerce-ios/assets/266376/4f2340dd-781b-4a76-9de3-1d1d93b09fe1

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift [hunk 1]
```diff
@@ -77,13 +76,13 @@ public enum BottomSheetProductType: Hashable, Identifiable {
         case .simple(let isVirtual):
             if isVirtual {
                 return NSLocalizedString(
-                    "bottomSheetProductType.simpleVirtual.title",
-                    value: "Simple virtual product",
+                    "bottomSheetProductType.simpleVirtualProduct.title",
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift:79]: Have to also change the key for the string value change to apply. Not completely sure why, but without doing this, the string change doesn't appear in the simulator.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift [hunk 2]
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
...
+        ForEach(groupedProductTypes, id: \.0) { (category, productTypes) in
+            VStack {
+                Text(category.label)
+                    .subheadlineStyle()
+                    .textCase(.uppercase)
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                    .padding(.bottom, Constants.categoryVerticalSpacing)
+                    .padding(.horizontal, Constants.horizontalSpacing)
+
+                ForEach(productTypes, id: \.self) { productType in
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: Nit: `BottomSheetProductType` is already conforming to `Identifiable`, so we can omit the `id` parameter here.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift [hunk 3]
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
...
-                VStack(alignment: .leading, spacing: Constants.verticalSpacing) {
-                    Text(productType.actionSheetTitle)
-                        .bodyStyle()
-                    Text(productType.actionSheetDescription)
-                        .subheadlineStyle()
+        ForEach(groupedProductTypes, id: \.0) { (category, productTypes) in
+            VStack {
+                Text(category.label)
+                    .subheadlineStyle()
+                    .textCase(.uppercase)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:29]: TIL about the `.textCase` modifier. Normally I would simply use `Text(category.label.uppercased)` to transform the string directly.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift [hunk 4]
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
...
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                    .padding(.bottom, Constants.categoryVerticalSpacing)
+                    .padding(.horizontal, Constants.horizontalSpacing)
+
+                ForEach(productTypes, id: \.self) { productType in
+                    HStack(alignment: .top, spacing: Constants.margin) {
+                        Image(uiImage: productType.actionSheetImage.withRenderingMode(.alwaysTemplate))
+                            .font(.title3)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: The modifier `font` will work only with system images. With custom assets, you might need to set the size explicitly with the`frame` modifier. To make it scalable, you'd need to get the scale from `ScaleMetric` property wrapper.



As you see above, the custom icons are not scaled in large font sizes.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift [hunk 5]
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
...
-                onOptionSelected(productType)
             }
         }
     }
 }
 
 private extension ManualProductTypeOptions {
     enum Constants {
+        // List of product categories. The ordering dictates how the categories are displayed.
+        static let productCategoriesOrder: [BottomSheetProductType.ProductCreationCategory] = [.standard, .subscription, .other]
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: Suggestion: You can omit this if you make `BottomSheetProductType.ProductCreationCategory` conform to `CaseIterable`. Then you can get the list of cases in the order declared in the enum with `BottomSheetProductType.ProductCreationCategory.allCases`.

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift [hunk 6]
```diff
@@ -199,6 +198,19 @@ public enum BottomSheetProductType: Hashable, Identifiable {
...
+    var productCreationCategory: ProductCreationCategory {
+        switch self {
+        case .simple, .blank, .variable:
+            return .standard
+        case .subscription, .variableSubscription:
+            return .subscription
+        case .custom, .grouped, .affiliate:
+            return .other
+        }
+    }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift]: Suggestion: you can avoid having to filter and match the category types if you replace this with a helper variable in `ProductCreationCategory` instead:

```Swift
extension ProductCreationCategory {
    var productTypes: BottomSheetProductType {
        // return the respective types for each category.
    }
}
```

Then in `ManualProductTypeOptions` you can get rid of `groupedProductTypes` and build the view by directly traversing the categories:

```Swift
ForEach(BottomSheetProductType.ProductCreationCategory.allCases, id: \.self) { category in
     ForEach(category.productTypes) { type in
          // TODO
     }
}
```

**Review by itsmeichigo — APPROVED** [review]
The update looks good! I wonder if we should also replace the bottom sheet for the case where AI is not eligible to use the same sheet but without the AI option. Currently, when testing for such sites, I still see the old sheet. Please feel free to file an issue for this as an enhancement separately though.



I left a few suggestions for code improvements below, but I'm pre-approving this PR.

**Comment by rachelmcr** [comment]
@hafizrahman Is this PR being actively worked on? Just checking because it has been moved to a future milestone a handful of times already. (I'll go ahead and move it to the next milestone for now.)

**Comment by hafizrahman** [comment]
@rachelmcr this was a hack week project that I didn't end up having enough time to complete, then got deprioritized over other tasks. Thanks for the reminder, I'll set as draft for now and come back to it next hack week.

**Comment by itsmeichigo** [comment]
@hafizrahman It would be great to wrap this PR for this HACK week as well, it has been around for quite a while.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift (hunk 2 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: 👍🏼  updated in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/34442d134d84058e62685dcfb68a6ecdfc1bae34

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift (hunk 4 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: Updated in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/34442d134d84058e62685dcfb68a6ecdfc1bae34

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift (hunk 6 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift]: I did some refactor on https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/f1b37ed71271104480c29d1be2724a266c109d5b based on your ideas here, but it doesn't fully implement your suggestion as I understood it.

The main thing is that the `ManualProductTypeOptions` receives `supportedProductTypes` that contains only the supported product types (calculated in  `presentActionSheetWithAI()` in `AddProductCoordinator`), so filtering things based on it is inevitable.

File: WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift (hunk 5 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/f1b37ed71271104480c29d1be2724a266c109d5b  also this comment is related to this https://github.com/woocommerce/woocommerce-ios/pull/12273#discussion_r1634539342

**Pull request merged**